src/sympc/protocol/aby3/aby3.py

```diff
@@ -277,22 +294,133 @@ def bit_injection(x: MPCTensor, session: Session, ring_size: int) -> MPCTensor:
+        result: List[MPCTensor] = []
+        for idx in range(ring_bits):
+            s = a[idx] + b[idx] + c[idx]
+            c[idx + 1] = a[idx] * b[idx] + c[idx] * (a[idx] + b[idx])
+            result.append(s)
+        return result
+
+    @staticmethod
+    def bit_decomposition(x: MPCTensor, session: Session) -> List[MPCTensor]:
+        """Perform ABY3 bit decomposition for conversion of arithmetic share to binary share.
+
+        Args:
+            x (MPCTensor): Arithmetic shares of secret.
+            session (Session): session the share belongs to.
+
+        Returns:
+            bin_share (List[MPCTensor]): Returns binary shares of each bit of the secret.
+
+        TODO : Should be modified to use parallel prefix adder when multiprocessing
+        functionality is integrated,currently unused.
+        """
+        ring_size = session.ring_size
+        ring_bits = get_nr_bits(ring_size)
+        x1: List[MPCTensor] = []  # bit shares of shares
+        x2: List[MPCTensor] = []
+        x3: List[MPCTensor] = []
+
+        args = [[share, str(2), True] for share in x.share_ptrs]
+
+        decompose = parallel_execution(ABY3.local_decomposition, session.parties)(args)
+
+        x_sh = list(zip(*decompose))
+
+        for idx in range(ring_bits):
+            x1_sh, x2_sh, x3_sh = zip(*x_sh[idx])
+            x1_sh = [ptr.resolve_pointer_type() for ptr in x1_sh]
+            x2_sh = [ptr.resolve_pointer_type() for ptr in x2_sh]
+            x3_sh = [ptr.resolve_pointer_type() for ptr in x3_sh]
+
+            x1_m = MPCTensor(shares=x1_sh, session=session, shape=x.shape)
+            x2_m = MPCTensor(shares=x2_sh, session=session, shape=x.shape)
+            x3_m = MPCTensor(shares=x3_sh, session=session, shape=x.shape)
+
+            x1.append(x1_m)
+            x2.append(x2_m)
+            x3.append(x3_m)
+
+        x1_2 = ABY3.full_adder(x1, x2, session)
+        bin_share = ABY3.full_adder(x1_2, x3, session)
+
+        return bin_share
+
+    @staticmethod
+    def bit_decomposition_ttp(x: MPCTensor, session: Session) -> List[MPCTensor]:
+        """Perform ABY3 bit decomposition using orchestrator as ttp.
+
+        Args:
+            x (MPCTensor): Arithmetic shares of secret.
+            session (Session): session the share belongs to.
+
+        Returns:
+            b_sh (List[MPCTensor]): Returns binary shares of each bit of the secret.
+
+        TODO: We should modify to use parallel prefix adder, which requires multiprocessing.
+        """
+        # Decoding is not done as they are shares of PRRS.
+        tensor = x.reconstruct(decode=False)
+        b_sh: List[MPCTensor] = []  # binary shares of bits
+        ring_size = session.ring_size
+        shares_sum = ReplicatedSharedTensor.shares_sum
+        ring_bits = get_nr_bits(ring_size)
+
+        for idx in range(ring_bits):
+            bit_mask = torch.ones(tensor.shape, dtype=tensor.dtype) << idx
```

Comment:
What does this do?

Reply:
We place 1 at specified position to extract the bit  by multiplying.

Reply:
For example to extract one at location 5 ,we place one at 5 bit pos and do bitwise and to extract